syntax = "proto3";

// The backup service definition.
service Backup {
  // Setup the backup service.
  rpc Setup (SetupRequest) returns (SetupResponse);

  // Sends a block of data to be written.
  rpc WriteBlock (stream WriteBlockRequest) returns (WriteBlockResponse);

  // Reads a block of data.
  rpc ReadBlock (stream ReadBlockRequest) returns (stream ReadBlockResponse);
}

message SetupRequest {
  uint64 size = 2; // The size of each block
}

message SetupResponse {
  bool success = 1; // Indicates if the setup was successful
  string message = 2; // Additional information or error message
}

// The request message containing the data to be written.
message WriteBlockRequest {
  uint64 block_no = 1; // The block number to write to
  bytes data = 2; // The data to write
}

// The response message for write requests.
message WriteBlockResponse {
  bool success = 1; // Indicates if the write was successful
  string message = 2; // Additional information or error message
}

// The request message for reading a block.
message ReadBlockRequest {
  uint64 block_no = 1; // The block number to read from
}

// The response message for read requests.
message ReadBlockResponse {
  bool success = 1; // Indicates if the read was successful
  string message = 2; // Additional information or error message
  bytes data = 3; // The data that was read
}
